### PR TITLE
fix(internal): add locks for thread-safety of trace encoder

### PR DESCRIPTION
**Pull request by majorgreys**
## Commit Message

{{title}}

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by brettlangdon — COMMENTED** [review]
we also need a lock around `flush()` (which is why we need the re-entrant lock)

File: tests/tracer/test_encoders.py [hunk 1]
```diff
@@ -595,3 +598,71 @@ def test_encoding_invalid_data(data):
         encoder.put(trace)
 
     assert encoder.encode() is None
+
+
+def test_custom_msgpack_encode_multithreaded():
```
Comment (brettlangdon) [inline tests/tracer/test_encoders.py:603]: while this test confirms the behavior and issue we are seeing, it is very dependent on implementation details which can change at any time. For example, the lock is currently in the wrong place, but if it was moved but still wrong, then this test might not fail or become flaky.

the maintenance burden of this test case is likely going to be high, and we'll probably miss a real failure thinking it is just flaky.

Comment (P403n1x87) [reply tests/tracer/test_encoders.py:603]: Yeh apologies for not spotting this earlier. I just happened to realise the lock mess once I moved away from the code and looked it at it recently with different 👀 😞 . I haven't had the chance to look closer to the details but I believe that if we guard all the encoder state that is shared by threads we should be good. The state includes, at the very least, the actual byte buffer and the trace count (same for the string table).

**Comment by majorgreys** [comment]
> we also need a lock around `flush()` (which is why we need the re-entrant lock)

I was thinking of putting the lock around `encode()` (and `put()` which are the interface to `BufferedEncoder`. @brettlangdon What do you think?

**Comment by brettlangdon** [comment]
> > we also need a lock around `flush()` (which is why we need the re-entrant lock)
> 
> I was thinking of putting the lock around `encode()` (and `put()` which are the interface to `BufferedEncoder`. @brettlangdon What do you think?

@majorgreys I am unsure the best strategy here, do we just guard around the top level/public API for these classes with Locks or do we put guards around _any_ interactions with the msgpack buffer/sizes/lengths/etc with RLocks?

RLocks aren't going to be as performant, but might be splitting hairs for safety/maintainability?

**Comment by majorgreys** [comment]
We discussed the strategy for locking offline and agreed that we want to go with reentrant locks for safety for now and wrap the API (essentially the `cpdefs` for the encoder cython classes) with locks. We will review the performance impact of using reentrant locks as a follow up.

The test case here will likely be flaky but the intention is that it fails at least once if there is some race condition on the buffer across any of the test runs. Another test which attempts to reproduce a bug is available at https://gist.github.com/majorgreys/0842496f988d6b1adbc7d184fb1ee6bb however it depends on implementation details of the encoder as it is which would make adding such a test a problem for maintenance and confidence.

**Review by Kyle-Verhoog — APPROVED** [review]

**Comment by brettlangdon** [comment]
@Mergifyio backport 0.56

File: ddtrace/internal/_encoding.pyx [hunk 2]
```diff
@@ -284,10 +284,11 @@ cdef class MsgpackStringTable(StringTable):
         self._sp_len = 0
 
     cpdef flush(self):
-        try:
-            return self.get_bytes()
-        finally:
-            self.reset()
+        with self._lock:
```
Comment (brettlangdon) [inline ddtrace/internal/_encoding.pyx:287]: should we do it around `def size` above as well?

File: ddtrace/internal/_encoding.pyx [hunk 3]
```diff
@@ -372,7 +373,7 @@ cdef class MsgpackEncoderBase(BufferedEncoder):
         self.max_size = max_size
         self.pk.buf_size = buf_size
         self.max_item_size = max_item_size if max_item_size < max_size else max_size
-        self._lock = threading.Lock()
+        self._lock = threading.RLock()
```
Comment (brettlangdon) [inline ddtrace/internal/_encoding.pyx:376]: Do we also need a lock around `ListBufferedEncoder.put`? (or well, move it?)

`encode_item` should probably also have a lock around it, but better to be sure it is around `put`?

File: ddtrace/internal/_encoding.pyx [hunk 4]
```diff
@@ -437,40 +439,40 @@ cdef class MsgpackEncoderBase(BufferedEncoder):
...
+                    raise BufferFull(self.size - size_before)
+
+                self._count += 1
+            except:
+                # rollback
+                self.pk.length = len_before
+                raise
 
     @property
     def size(self):
```
Comment (brettlangdon) [inline ddtrace/internal/_encoding.pyx:478]: do we want a lock here too?

**Review by brettlangdon — CHANGES_REQUESTED** [review]
we need additional locking around the v05 encoder as well

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**